### PR TITLE
Add eslint-react-compiler 

### PR DESCRIPTION
**Pull request by MrFlashAccount**
### Pull Request Description

This PR adds eslint rules for React compiler, to make our codebase ready once react-compiler is ready for landing

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

File: app/gui/src/dashboard/components/AriaComponents/Form/components/FormProvider.tsx [hunk 1]
```diff
@@ -39,6 +39,7 @@ export function useFormContext<Schema extends types.TSchema>(
   if (form != null && 'control' in form) {
     return form
   } else {
+    // eslint-disable-next-line react-compiler/react-compiler
```
Comment (somebody1234) [inline app/gui/src/dashboard/components/AriaComponents/Form/components/FormProvider.tsx:42]: i think we should add comments for all such lint disables. i assume this one is conditional calling given the `rules-of-hooks` ignore but still worth adding a comment imo.

(alternative we should each branch up into its own component?)

File: app/gui/src/dashboard/hooks/mountHooks.ts [hunk 2]
```diff
@@ -36,11 +36,12 @@ export function useMounted(callback: () => void) {
 
 /** Returns a function that returns `true` if the component renders for the first time. */
 export function useIsFirstRender() {
-  const isFirstMount = useRef(true)
+  // We use `null` here instead of `true` to make react-compiler happy.
```
Comment (somebody1234) [inline app/gui/src/dashboard/hooks/mountHooks.ts]: would be nice to explain why `react-compiler` does not like `true`

**Review by somebody1234 — APPROVED** [review]

File: app/gui/src/dashboard/components/AriaComponents/Form/components/FormProvider.tsx (hunk 1 above)
Comment (MrFlashAccount) [reply app/gui/src/dashboard/components/AriaComponents/Form/components/FormProvider.tsx:42]: well, there's a thing that it's safe to call `useContext` conditionally and `React` allows that, though it's not documented anywhere 🌚

In react@19 there'll be a `use` function that can read context values and doesn't violate rules of hooks when called conditionally.

File: app/gui/src/dashboard/hooks/mountHooks.ts (hunk 2 above)
Comment (MrFlashAccount) [reply app/gui/src/dashboard/hooks/mountHooks.ts]: Will attach a reference

File: app/gui/package.json [hunk 3]
```diff
@@ -21,7 +21,7 @@
     "build": "vite build",
     "build-cloud": "cross-env CLOUD_BUILD=true corepack pnpm run build",
     "preview": "vite preview",
-    "lint": "eslint . --max-warnings=0",
```
Comment (farmaazon) [inline app/gui/package.json:24]: Why this change? I want to fail on warnings.

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui/package.json (hunk 3 above)
Comment (somebody1234) [reply app/gui/package.json:24]: yeah, at the very least we should keep it for `project-view`

maybe for the time being we can split the lint command (back) into separate gui and dashboard lint commands filtered by top-level directory?

Comment (MrFlashAccount) [reply app/gui/package.json:24]: > I want to fail on warnings.

Why, though? If we want to fail on warnings, maybe it's better to switch those into errors instead?

We added eslint rules for react compiler and we don't want to suppress them everywhere we have violations(sinсe they are not mistakes, linter points out that the component with the warnings won't be optimized by the react-compiler). We want to see these places, but they shouldn't fail the lint task

If you don't want to see extra noise on CI, let's add the `--quiet` [flag](https://eslint.org/docs/latest/use/command-line-interface#--quiet)

Using `warn` flag for eslint-compiler complies with the reasons why `warn` exists: https://eslint.org/docs/latest/use/configure/rules#rule-severities

Comment (farmaazon) [reply app/gui/package.json:24]: I want to keep distinction between error requiring my immediate attention (because otherwise the program won't  work) from warnings (this is bad, but it may wait before merge). It just improves the development experience, when missing jsdocs won't make the function look like it was completely broken.

To keep everyone happy, maybe we could make the separate script which run linter with additional --plugin and --rule option, which will not be run on CI?

Comment (MrFlashAccount) [reply app/gui/package.json:24]: I think this makes the DX worse, because `warnings` are not supposed to break the task, see: https://eslint.org/docs/latest/use/configure/rules#rule-severities

Comment (farmaazon) [reply app/gui/package.json:24]: From call: we agreed that we keep warnings for react-compiler, but set here limit to their exact current number (so any new warning will still fail the CI) with a comment describing what warnings we want to leave.

**Review by farmaazon — APPROVED** [review]
Approving to not block it anymore, but see the comment.

File: app/gui/package.json (hunk 3 above)
Comment (MrFlashAccount) [reply app/gui/package.json:24]: @farmaazon addressed your concerns
